first cut at git-style zed lake (#2548)
brimdata/zed: docs/lake/design.md

The semantics of a Zed lake loosely follows the nomenclature and
design patterns of `git`.  In this approach,

> * a _lake_ is like a `github` organization,

Reviewer: ```suggestion
* a _lake_ is like a GitHub organization,
```

Note that there may be multiple pool keys, where subsequent keys act as the secondary,
tertiary, and so forth sort key.

> In all these examples, the lake identity is implied by its path (e.g., an s3 URI

Reviewer: ```suggestion
In all these examples, the lake identity is implied by its path (e.g., an S3 URI
```

or a file system path) and may be specified by the ZED_LAKE_ROOT environment variable

> when running the `zed lake` tooling on a local host.  In a cloud deployment

Reviewer: ```suggestion
when running `zed lake` commands on a local host.  In a cloud deployment
```

or running queries through an application, the lake path is determined by
an authenticated connection to the Zed lake service, which explicitly denotes

> the lake name (analagous to how a github user authenticates access to
> a named github organization).

Reviewer: ```suggestion
the lake name (analagous to how a GitHub user authenticates access to
a named GitHub organization).
```

These steps can be explicitly run in stages, e.g.,
```
zed lake add -p logs sample.json
(commit <tag> printed to stdout)
zed lake commit -p logs <tag>
```

> A commit `<tag>` refers to one or more data objects that is stored in the

Reviewer: ```suggestion
A commit `<tag>` refers to one or more data objects stored in the
```

Alternatively, a reader can scan a specific set of commits by enumerating
the commit tags in the scan/search API.

> Also, arbitrary meta-data can be committed to the log as described below,

Reviewer: ```suggestion
Also, arbitrary metadata can be committed to the log as described below,
```

> e.g., to associate index objects or derived analytics to specific in

Reviewer: I'm not sure what "to specific in" means here.

from potentially a large number of different segments.

To solve this problem, the Zed lake design follows the
the [LSM](https://en.wikipedia.org/wiki/Log-structured_merge-tree) design pattern.

> Since each segment is stored in sorted order, a total order over a collection

Reviewer: ```suggestion
Since records in each segment are stored in sorted order, a total order over a collection
```

To avoid consistency issues here, the old commits need
to be deleted while simultaneously adding the new commit.

> This can be done automicall
> by performing a merge, staging the deletes, then
> commit the add and merge together:

Reviewer: ```suggestion
This can be done atomically
by performing a merge, staging the deletes, and then
committing the merge and delete together:
```

```
zed lake merge -p logs <tag-1> <tag-2> <tag-3>
(merged commit <merge-tag> printed to stdout)
zed lake delete -p logs <tag-1> <tag-2> <tag-3>
(delete commit <del-tag> printed to stdout)
zed lake commit -p logs <merge-tag> <del-tag>
```
Note that the data in commits `<tag-1>`, `<tag-2>`, and `<tag-3>` remains
in the pool and scans can be performed on older snapshots of the pool
as long as the data is not deleted.

> Data can be delted with the DANGER-ZONE command `zed lake purge`.

Reviewer: ```suggestion
Data can be deleted with the DANGER-ZONE command `zed lake purge`.
```

the scan schedule can be quickly computed in a small number of round-trips
(that navigate very wide B-trees) to cloud object storage.

### Mutable Objects

> Mutable objects are built upon a journal of arbitrary set updates
> to a one ore more key-value entities stored within a commit journal.

Reviewer: ```suggestion
Mutable objects are built upon a journal of arbitrary updates
to sets of one ore more key-value entities stored within a commit journal.
```
